neo/ui-stm32/src/hal/clock.rs

```diff
@@ -0,0 +1,168 @@
+//! - `validate`: Validates the clock configuration to ensure it is set up correctly.
+//!
+//! ## Usage
+//!
+//! The `init` function should be called during system startup to configure the clock. The `validate` function
+//! can be used to check if the clock configuration is correct.
+//!
+
+use crate::svd::{Read, Write, FLASH, RCC};
+
+fn write_pll_m(pll_m: u8) {
+    assert!(pll_m < (1 << 6));
+    RCC::PLLCFGR::PLLM5::write((pll_m >> 5) & 0x1 == 1);
+    RCC::PLLCFGR::PLLM4::write((pll_m >> 4) & 0x1 == 1);
+    RCC::PLLCFGR::PLLM3::write((pll_m >> 3) & 0x1 == 1);
+    RCC::PLLCFGR::PLLM2::write((pll_m >> 2) & 0x1 == 1);
+    RCC::PLLCFGR::PLLM1::write((pll_m >> 1) & 0x1 == 1);
+    RCC::PLLCFGR::PLLM0::write((pll_m >> 0) & 0x1 == 1);
+}
+
+fn write_pll_n(pll_n: u8) {
+    RCC::PLLCFGR::PLLN7::write((pll_n >> 7) & 0x1 == 1);
+    RCC::PLLCFGR::PLLN6::write((pll_n >> 6) & 0x1 == 1);
+    RCC::PLLCFGR::PLLN5::write((pll_n >> 5) & 0x1 == 1);
+    RCC::PLLCFGR::PLLN4::write((pll_n >> 4) & 0x1 == 1);
+    RCC::PLLCFGR::PLLN3::write((pll_n >> 3) & 0x1 == 1);
+    RCC::PLLCFGR::PLLN2::write((pll_n >> 2) & 0x1 == 1);
+    RCC::PLLCFGR::PLLN1::write((pll_n >> 1) & 0x1 == 1);
+    RCC::PLLCFGR::PLLN0::write((pll_n >> 0) & 0x1 == 1);
+}
+
+fn write_pll_p(pll_p: u8) {
+    assert!(pll_p < (1 << 2));
+    RCC::PLLCFGR::PLLP1::write((pll_p >> 1) & 0x1 == 1);
+    RCC::PLLCFGR::PLLP0::write((pll_p >> 0) & 0x1 == 1);
+}
+
+fn write_pll_q(pll_q: u8) {
+    assert!(pll_q < (1 << 4));
+    RCC::PLLCFGR::PLLQ3::write((pll_q >> 3) & 0x1 == 1);
+    RCC::PLLCFGR::PLLQ2::write((pll_q >> 2) & 0x1 == 1);
+    RCC::PLLCFGR::PLLQ1::write((pll_q >> 1) & 0x1 == 1);
+    RCC::PLLCFGR::PLLQ0::write((pll_q >> 0) & 0x1 == 1);
+}
+
+fn read_pll_n() -> u8 {
+    (u8::from(RCC::PLLCFGR::PLLN7::read()) << 7)
+        | (u8::from(RCC::PLLCFGR::PLLN6::read()) << 6)
+        | (u8::from(RCC::PLLCFGR::PLLN5::read()) << 5)
+        | (u8::from(RCC::PLLCFGR::PLLN4::read()) << 4)
+        | (u8::from(RCC::PLLCFGR::PLLN3::read()) << 3)
+        | (u8::from(RCC::PLLCFGR::PLLN2::read()) << 2)
+        | (u8::from(RCC::PLLCFGR::PLLN1::read()) << 1)
+        | (u8::from(RCC::PLLCFGR::PLLN0::read()) << 0)
+}
+
+fn read_pll_p() -> u8 {
+    (u8::from(RCC::PLLCFGR::PLLP1::read()) << 1) | (u8::from(RCC::PLLCFGR::PLLP0::read()) << 0)
+}
+
+fn read_pll_q() -> u8 {
+    (u8::from(RCC::PLLCFGR::PLLQ3::read()) << 3)
+        | (u8::from(RCC::PLLCFGR::PLLQ2::read()) << 2)
+        | (u8::from(RCC::PLLCFGR::PLLQ1::read()) << 1)
+        | (u8::from(RCC::PLLCFGR::PLLQ0::read()) << 0)
+}
+
+/// Initializes the clock configuration based on the board-specific settings.
+pub fn init(hse_clk_freq: u32) {
+    let pll_m = match hse_clk_freq {
+        16_000_000 => 8,
+        24_000_000 => 12,
+        _ => unreachable!("Invalid HSE clock frequency {hse_clk_freq}"),
+    };
+
+    // XXX(rlb): These fields are read-only (?)
```

Comment:
These are not needed by the STM32F405  - if they are written it is more or lest ignored - but they were use by the emulator to set up initial state.